Add unit test for TriggerResults format

wddgit opened the pull request:
#### PR description:

Add a new unit test for the TriggerResults data format. This generates a data file containing a TriggerResults object with known content. Then it reads it. It verifies that when we read it we obtain values that match the known written values for all the data fields in the object. In particular all containers have content so all contained types are also read.

It also reads the old files in the DataFormats/Common data repository which are listed in the shell script. The plan is that each time the data format of TriggerResults is modified a file will added.

#### PR validation:

This only adds a unit test in DataFormats/Common which passes. It shouldn't affect anything else.

wddgit committed:
Add unit test for TriggerResults format

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

- DataFormats/Common (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

wddgit commented:
A couple of comments on related history:

The TriggerResults format has not changed since 2006. In 2014 something in ROOT changed that affected the checksum and we had to add a new version number in classes_def.html.

We have an existing unit test that is similar to this new one. It is in IOPool/Input/test/TestPoolInput.sh. See this line and what follows it: "for file in ${IOPoolInputData}/IOPool/Input/data/raw*.root". It reads an old existing file with RAW and then writes a new file. A second process reads the new converted file. There is not an analyzer that reads content, just PoolSource and PoolOutputModule. Here is the content of the old file:

Singularity> edmFileUtil file:/cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_1_X_2023-04-18-1100/external/el8_amd64_gcc11/data/IOPool/Input/data/rawData_old_format_CMSSW_4_2_8.root -P
file:/cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_1_X_2023-04-18-1100/external/el8_amd64_gcc11/data/IOPool/Input/data/rawData_old_format_CMSSW_4_2_8.root
file:/cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_1_X_2023-04-18-1100/external/el8_amd64_gcc11/data/IOPool/Input/data/rawData_old_format_CMSSW_4_2_8.root (1 runs, 1 lumis, 1 events, 763640 bytes)
Branch 0 of Events tree: EventAuxiliary Total size = 757 Compressed size = 232
Branch 1 of Events tree: EventBranchEntryInfo Total size = 898 Compressed size = 209
Branch 2 of Events tree: EventSelections Total size = 681 Compressed size = 152
Branch 3 of Events tree: BranchListIndexes Total size = 647 Compressed size = 110
Branch 4 of Events tree: FEDRawDataCollection_source__HLT. Total size = 372335 Compressed size = 137624
Branch 5 of Events tree: L1GlobalTriggerObjectMapRecord_hltL1GtObjectMap__HLT. Total size = 24654 Compressed size = 3622
Branch 6 of Events tree: edmTriggerResults_TriggerResults__HLT. Total size = 6183 Compressed size = 640
Branch 7 of Events tree: triggerTriggerEvent_hltTriggerSummaryAOD__HLT. Total size = 15629 Compressed size = 4894

I think the intent was that new versions of this file would be added if RAW formats changed. That never happened, there is only the CMSSW_4_2_8 file. Although only the RAW format L1GlobalTriggerObjectMapRecord changed. It is a completely different type now apparently. The other formats have not changed.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-503e27/32113/summary.html
**COMMIT**: 06b1d2dd23e293fc9f1aaed403d1200d2114378c
**CMSSW**: CMSSW_13_1_X_2023-04-24-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41395/32113/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 8 lines to the logs
 - Reco comparison results: 12 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460685
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460657
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-24-1100+503e27/56644/triggerResults): no differences found

makortel reviewed (commented):
Thanks @wddgit, overall looks good. Could you also create a `DataFormats/Common/README.md` with content along

> # DataFormats/Common
>
> ## `edm::TriggerResults`
>
> The `edm::TriggerResults` is part of the RAW data, and any changes must be backwards compatible. In order to ensure it can be read by all future CMSSW releases, there is a `TestTriggerResultsFormat` unit test, which makes use of `TestReadTriggerResults` analyzer and `TestWriteTriggerResults` producer. The unit test checks that the object can be read properly from
> * a file in the same release as it was written
> * files written by (some) earlier releases can be read
>
> If the `edm::TriggerResults` gets changed in the future, please adjust the `TestReadTriggerResults` and `TestWriteTriggerResults` modules accordingly. It is important that every member container has some content in this test. Please also add a new file to https://github.com/cms-data/DataFormats-Common/ repository, and update the `TestTriggerResultsFormat` unit test to read the newly created file. The file name should contain the release or pre-release with which it was written. 

@Dr15Jones Do you have comments on the README contents?

smuzaffar commented:
@wddgit , yes just open a PR for that repo then just test this PR along with cms-data one e.g. `please test with cms-data/DataFormats-Common#1` should be enough. As @makortel [wrote](https://github.com/cms-sw/cmssw/pull/41395#discussion_r1176676911) you can use `edmFileInPath DataFormats/Common/data/<filename>` to get the full path of the data file

wddgit commented on DataFormats/Common/test/TestTriggerResultsFormat.sh:
```diff
@@ -0,0 +1,26 @@
+for file in ${DataFormatsCommonData}/DataFormats/Common/data/oldTestTriggerResults*.root
+do
+  cmsRun ${LOCAL_TEST_DIR}/test_readTriggerResults_cfg.py "$file" || die "Failed to read old file $file" $?
+done
```
Done. That is better.

wddgit commented:
I think the commit I just pushed addresses all the comments. I squashed it.

The shell script changed significantly. The data file was removed. The documentation file was added. The rest of it remained exactly the same as before.

smuzaffar commented:
test parameters:
- pull_request = cms-data/DataFormats-Common#1

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41395/35273

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #41395 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented:
please test with cms-data/DataFormats-Common#1

Note the unit test will fail unless merged at the same time as or after the data file PR.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-503e27/32143/summary.html
**COMMIT**: cc027c6e4ab4c0c6343822639f95b7c70232ff3a
**CMSSW**: CMSSW_13_1_X_2023-04-25-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41395/32143/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @aloeliger cms-sw/cmssw#41401
 - @jfernan2 cms-sw/cmssw#41375
 - @iarspider cms-sw/cmssw#41377
 - @hahnjo cms-sw/cmssw#41392
 - @mmusich cms-sw/cmssw#41393
 - @menglu21 cms-sw/cmssw#41325
 - @mmusich cms-sw/cmssw#41378
 - @bsunanda cms-sw/cmssw#41390
 - @cms-sw cms-sw/cmsdist#8447
 - @cms-sw cms-sw/cmsdist#8466
 - @cms-sw cms-sw/cmsdist#8451

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-503e27/32143/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-503e27/32143/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 6 lines to the logs
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460685
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460657
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-25-1100+503e27/56669/triggerResults): no differences found

wddgit commented:
Just for the record, I wanted to make the comment that the new test added by this PR will not be very good at detecting the problem discussed in issues #41348 and #41246. For it to detect that particular problem, a test input file would have to be written using a release with that problem (note that particular problem was associated with an internal ROOT change and not a data format change) and then later (potentially years later) if the persistent format of TriggerResults changed, then this test would detect the problem.

That said, that problem motivated us to think about RAW data format testing and implement the improvements in this PR. Given that we are supposed to guarantee backward compatibility for RAW data formats forever and the importance of RAW data, the new test in this PR is a logical unit test to have and run.

makortel commented:
+core

Needs https://github.com/cms-data/DataFormats-Common/pull/1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
- https://github.com/cms-sw/cmsdist/pull/8479 was already merged
